nixos/modules/services/web-servers/nginx/default.nix

```diff
@@ -1209,10 +1211,10 @@ in
         '';
       }
     ] ++ map (name: mkCertOwnershipAssertion {
-      inherit (cfg) group user;
       cert = config.security.acme.certs.${name};
       groups = config.users.groups;
-    }) dependentCertNames;
+      services = [ config.systemd.services.nginx ] ++ lib.optional (cfg.enableReload || vhostCertNames != []) config.systemd.services.nginx-config-reload;
```

Comment:
nginx-config-reload service doesn't have serviceConfig.User, resolving in `error: attribute 'User' missing`

Reply:
Please provide more context (stack trace at the minimum).

I believe it's failing when formatting the assertion message, so workaround would be to fix the access. I can't say how from just this single line though.  
I'll fix the message if it's indeed that.

Reply:
I'd be really surprised if this is caused by a missing attribute, the test suite would have failed. Did the tests run on this PR?

Reply:
Yes, and I ran them locally.

Is there a test that verifies the assert fails though? And importantly here, that it fails with the expected message?
That's not something I looked into, but I can add it as part of the fix.

Reply:
I don't get this error when the assertions are satifised, only when they are not. I can provide the stack trace later.

Reply:
```
       … while calling anonymous lambda
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/attrsets.nix:1570:24:
         1569|     let f = attrPath:
         1570|       zipAttrsWith (n: values:
             |                        ^
         1571|         let here = attrPath ++ [n]; in

       … while calling the 'head' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/attrsets.nix:1574:11:
         1573|         || pred here (elemAt values 1) (head values) then
         1574|           head values
             |           ^
         1575|         else

       … while calling anonymous lambda
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/attrsets.nix:1204:18:
         1203|         mapAttrs
         1204|           (name: value:
             |                  ^
         1205|             if isAttrs value && cond value

       … from call site
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/attrsets.nix:1207:18:
         1206|             then recurse (path ++ [ name ]) value
         1207|             else f (path ++ [ name ]) value);
             |                  ^
         1208|     in

       … while calling anonymous lambda
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:254:72:
          253|           # For definitions that have an associated option
          254|           declaredConfig = mapAttrsRecursiveCond (v: ! isOption v) (_: v: v.value) options;
             |                                                                        ^
          255|

       … while evaluating the attribute 'value'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:816:9:
          815|     in warnDeprecation opt //
          816|       { value = addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          817|         inherit (res.defsFinal') highestPrio;

       … while evaluating the option `system.build.toplevel':

       … while evaluating the attribute 'mergedValue'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:851:5:
          850|     # Type-check the remaining definitions, and merge them. Or throw if no definitions.
          851|     mergedValue =
             |     ^
          852|       if isDefined then

       … while evaluating a branch condition
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:852:7:
          851|     mergedValue =
          852|       if isDefined then
             |       ^
          853|         if all (def: type.check def.value) defsFinal then type.merge loc defsFinal

       … while evaluating the attribute 'values'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:845:9:
          844|       in {
          845|         values = defs''';
             |         ^
          846|         inherit (defs'') highestPrio;

       … while evaluating a branch condition
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:841:11:
          840|           # Avoid sorting if we don't have to.
          841|           if any (def: def.value._type or "" == "order") defs''.values
             |           ^
          842|           then sortProperties defs''.values

       … while calling the 'any' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:841:14:
          840|           # Avoid sorting if we don't have to.
          841|           if any (def: def.value._type or "" == "order") defs''.values
             |              ^
          842|           then sortProperties defs''.values

       … while evaluating the attribute 'values'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:944:7:
          943|     in {
          944|       values = concatMap (def: if getPrio def == highestPrio then [(strip def)] else []) defs;
             |       ^
          945|       inherit highestPrio;

       … while calling the 'concatMap' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:944:16:
          943|     in {
          944|       values = concatMap (def: if getPrio def == highestPrio then [(strip def)] else []) defs;
             |                ^
          945|       inherit highestPrio;

       … while calling the 'concatMap' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:831:17:
          830|         # Process mkMerge and mkIf properties.
          831|         defs' = concatMap (m:
             |                 ^
          832|           map (value: { inherit (m) file; inherit value; }) (addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … while calling anonymous lambda
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:831:28:
          830|         # Process mkMerge and mkIf properties.
          831|         defs' = concatMap (m:
             |                            ^
          832|           map (value: { inherit (m) file; inherit value; }) (addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))

       … while calling the 'map' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:832:11:
          831|         defs' = concatMap (m:
          832|           map (value: { inherit (m) file; inherit value; }) (addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))
             |           ^
          833|         ) defs;

       … while evaluating definitions from `/nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/system/activation/top-level.nix':

       … from call site
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:832:128:
          831|         defs' = concatMap (m:
          832|           map (value: { inherit (m) file; inherit value; }) (addErrorContext "while evaluating definitions from `${m.file}':" (dischargeProperties m.value))
             |                                                                                                                                ^
          833|         ) defs;

       … while calling 'dischargeProperties'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:903:25:
          902|   */
          903|   dischargeProperties = def:
             |                         ^
          904|     if def._type or "" == "merge" then

       … while evaluating a branch condition
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:904:5:
          903|   dischargeProperties = def:
          904|     if def._type or "" == "merge" then
             |     ^
          905|       concatMap dischargeProperties def.contents

       … while evaluating the attribute 'value'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/lib/modules.nix:614:53:
          613|                 (n: value:
          614|                   [{ inherit (module) file; inherit value; }]
             |                                                     ^
          615|                 )

       … while calling the 'throw' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/system/activation/top-level.nix:73:10:
           72|   baseSystemAssertWarn = if failedAssertions != []
           73|     then throw "\nFailed assertions:\n${concatStringsSep "\n" (map (x: "- ${x}") failedAssertions)}"
             |          ^
           74|     else showWarnings config.warnings baseSystem;

       … while calling the 'concatStringsSep' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/system/activation/top-level.nix:73:41:
           72|   baseSystemAssertWarn = if failedAssertions != []
           73|     then throw "\nFailed assertions:\n${concatStringsSep "\n" (map (x: "- ${x}") failedAssertions)}"
             |                                         ^
           74|     else showWarnings config.warnings baseSystem;

       … while calling anonymous lambda
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/system/activation/top-level.nix:73:69:
           72|   baseSystemAssertWarn = if failedAssertions != []
           73|     then throw "\nFailed assertions:\n${concatStringsSep "\n" (map (x: "- ${x}") failedAssertions)}"
             |                                                                     ^
           74|     else showWarnings config.warnings baseSystem;

       … from call site
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/system/activation/top-level.nix:73:77:
           72|   baseSystemAssertWarn = if failedAssertions != []
           73|     then throw "\nFailed assertions:\n${concatStringsSep "\n" (map (x: "- ${x}") failedAssertions)}"
             |                                                                             ^
           74|     else showWarnings config.warnings baseSystem;

       … while calling anonymous lambda
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/system/activation/top-level.nix:70:27:
           69|
           70|   failedAssertions = map (x: x.message) (filter (x: !x.assertion) config.assertions);
             |                           ^
           71|

       … while evaluating the attribute 'message'
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/security/acme/mk-cert-ownership-assertion.nix:18:3:
           17|
           18|   message = "Certificate ${cert.domain} (group=${cert.group}) must be readable by service(s) ${
             |   ^
           19|     catSep ", " (map (svc: "${svc.name} (user=${svc.serviceConfig.User} groups=${catSep " " (svcGroups svc)})") services)

       … while calling the 'concatStringsSep' builtin
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/security/acme/mk-cert-ownership-assertion.nix:19:5:
           18|   message = "Certificate ${cert.domain} (group=${cert.group}) must be readable by service(s) ${
           19|     catSep ", " (map (svc: "${svc.name} (user=${svc.serviceConfig.User} groups=${catSep " " (svcGroups svc)})") services)
             |     ^
           20|   }";

       … while calling anonymous lambda
         at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/security/acme/mk-cert-ownership-assertion.nix:19:23:
           18|   message = "Certificate ${cert.domain} (group=${cert.group}) must be readable by service(s) ${
           19|     catSep ", " (map (svc: "${svc.name} (user=${svc.serviceConfig.User} groups=${catSep " " (svcGroups svc)})") services)
             |                       ^
           20|   }";

       error: attribute 'User' missing
       at /nix/store/n1g84klfb0h3bpwyvc59lcy5ca58h36w-source/nixos/modules/security/acme/mk-cert-ownership-assertion.nix:19:49:
           18|   message = "Certificate ${cert.domain} (group=${cert.group}) must be readable by service(s) ${
           19|     catSep ", " (map (svc: "${svc.name} (user=${svc.serviceConfig.User} groups=${catSep " " (svcGroups svc)})") services)
             |                                                 ^
           20|   }";
```

Reply:
Great thank you for the confirmation!

Reply:
Just ran into this as well. It's caused by `systemd.services.nginx-config-reload` not having a `serviceConfig.User` set here: https://github.com/NixOS/nixpkgs/blob/43f6a895e5169a1a5d35a7141e016ccfc828a28f/nixos/modules/services/web-servers/nginx/default.nix#L1322-L1343

The assertion failed in the first place because nginx didn't set the cert's group automatically. Normally, nginx sets the `group` of the certs it uses to its own `cfg.group`, but this only happens for certs "fully-managed" by nginx: https://github.com/NixOS/nixpkgs/blob/23e89b7da85c3640bbc2173fe04f4bd114342367/nixos/modules/services/web-servers/nginx/default.nix#L1357-L1358
The alternative (in order to use a single wildcard cert) is to set `virtualHosts.*.useACMEHost = "example.com";` instead, but that _doesn't_ trigger this update. To fix this and just do what nginx does to other certs, set `security.acme.certs."example.com".group = config.services.nginx.group;`.

Reply:
@ThinkChaos Taking a step back from the assertion/attribute error itself, what is the value in applying mkCertOwnershipAssertion to nginx-config-reload.service? It does not interact with the certs explicitly, only nginx itself does.